File: frontend/app/actions/workspaces.ts
"use server";

import { revalidatePath } from "next/cache";
import { apiFetch } from "@/lib/api";
import type { DepartmentItem, WorkspaceItem } from "@/lib/types";
import { responseErrorMessage } from "./errors";

export async function listWorkspaces(): Promise<WorkspaceItem[]> {
  const res = await apiFetch("/admin/v1/workspaces");
  if (!res.ok) throw new Error(`Failed to load workspaces (${res.status})`);
  return res.json() as Promise<WorkspaceItem[]>;
}

export async function listAllDepartments(): Promise<DepartmentItem[]> {
  const res = await apiFetch("/admin/v1/departments");
  if (!res.ok) throw new Error(`Failed to load departments (${res.status})`);
  return res.json() as Promise<DepartmentItem[]>;
}

export async function createWorkspace(
  name: string,
): Promise<{ ok: true; workspace: WorkspaceItem } | { ok: false; error: string }> {
  let res: Response;
  try {
    res = await apiFetch("/admin/v1/workspaces", {
      method: "POST",
      body: JSON.stringify({ name }),
    });
  } catch (e) {
    return { ok: false, error: (e as Error).message };
  }

  if (res.status === 409) {
    return { ok: false, error: "A workspace with that name already exists." };
  }

  if (!res.ok) {
    return { ok: false, error: await responseErrorMessage(res, `Create failed (${res.status})`) };
  }

  const workspace = (await res.json()) as WorkspaceItem;
  revalidatePath("/dashboard/workspaces");
  return { ok: true, workspace };
}

export async function deleteWorkspace(
  slug: string,
): Promise<{ ok: true } | { ok: false; error: string }> {
  let res: Response;
  try {
    res = await apiFetch(`/admin/v1/workspaces/${encodeURIComponent(slug)}`, {
      method: "DELETE",
    });
  } catch (e) {
    return { ok: false, error: (e as Error).message };
  }

  if (!res.ok) {
    return { ok: false, error: await responseErrorMessage(res, `Delete failed (${res.status})`) };
  }

  revalidatePath("/dashboard/workspaces");
  return { ok: true };
}

File: server/app/routers/admin/workspaces.py
from fastapi import APIRouter, HTTPException, status

from app.schemas.admin.workspaces import WorkspaceCreate, WorkspaceDeleteResponse, WorkspaceItem, WorkspaceUpdate
from app.services import admin_service

router = APIRouter()


@router.get("/workspaces", response_model=list[WorkspaceItem])
def list_workspaces():
    return admin_service.list_workspaces()


@router.post("/workspaces", response_model=WorkspaceItem, status_code=status.HTTP_201_CREATED)
def create_workspace(body: WorkspaceCreate):
    try:
        return admin_service.create_workspace(body.name)
    except admin_service.DuplicateSlug as exc:
        raise HTTPException(status_code=409, detail="Workspace slug already exists") from exc


@router.patch("/workspaces/{slug}", response_model=WorkspaceItem)
def rename_workspace(slug: str, body: WorkspaceUpdate):
    try:
        return admin_service.rename_workspace(slug, body.name)
    except admin_service.WorkspaceNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.delete("/workspaces/{slug}", response_model=WorkspaceDeleteResponse)
def delete_workspace(slug: str):
    try:
        return admin_service.delete_workspace(slug)
    except admin_service.WorkspaceNotFound as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
